File: android-app/feature/feature-social/src/main/java/tv/bayit/plus/feature/social/chess/ChessViewModel+Actions.kt
package tv.bayit.plus.feature.social.chess

import tv.bayit.plus.core.common.BayitResult

private const val DEEP_LINK_CHESS_BASE = "bayitplus://chess"

fun ChessViewModel.createGameForWhatsApp(color: String, timeControl: Int?) {
    launchInScope {
        _uiState.value = ChessUiState.Loading
        logger.info("Creating WhatsApp challenge game", mapOf("color" to color))
        when (val result = chessRepository.createGame(color, "pvp", null, timeControl)) {
            is BayitResult.Success -> {
                val game = result.data
                val link = "$DEEP_LINK_CHESS_BASE/${game.gameCode}"
                val message = stringProvider.string(
                    "chess.whatsAppMessage",
                    mapOf("code" to game.gameCode, "link" to link),
                )
                setPendingWhatsAppMessage(message)
                transitionToGame(game)
            }
            is BayitResult.Error -> {
                logger.error("WhatsApp challenge create failed", result.exception)
                _uiState.value = ChessUiState.Error(
                    result.message ?: result.exception.message.orEmpty()
                )
            }
            is BayitResult.Loading -> Unit
        }
    }
}

fun ChessViewModel.createGame(
    color: String,
    gameMode: String,
    botDifficulty: String?,
    timeControl: Int?,
) {
    launchInScope {
        _uiState.value = ChessUiState.Loading
        logger.info("Creating chess game", mapOf("color" to color, "mode" to gameMode))
        when (val result = chessRepository.createGame(color, gameMode, botDifficulty, timeControl)) {
            is BayitResult.Success -> transitionToGame(result.data)
            is BayitResult.Error -> {
                logger.error("Create game failed", result.exception)
                _uiState.value = ChessUiState.Error(
                    result.message ?: result.exception.message.orEmpty()
                )
            }
            is BayitResult.Loading -> Unit
        }
    }
}

fun ChessViewModel.joinGame(gameCode: String) {
    launchInScope {
        _uiState.value = ChessUiState.Loading
        logger.info("Joining chess game", mapOf("gameCode" to gameCode))
        when (val result = chessRepository.joinGame(gameCode)) {
            is BayitResult.Success -> transitionToGame(result.data)
            is BayitResult.Error -> {
                logger.error("Join game failed", result.exception)
                _uiState.value = ChessUiState.Error(
                    result.message ?: result.exception.message.orEmpty()
                )
            }
            is BayitResult.Loading -> Unit
        }
    }
}

fun ChessViewModel.loadGame(gameCode: String) {
    launchInScope {
        _uiState.value = ChessUiState.Loading
        logger.info("Loading chess game", mapOf("gameCode" to gameCode))
        when (val result = chessRepository.getGame(gameCode)) {
            is BayitResult.Success -> transitionToGame(result.data)
            is BayitResult.Error -> {
                logger.error("Load game failed", result.exception)
                _uiState.value = ChessUiState.Error(
                    result.message ?: result.exception.message.orEmpty()
                )
            }
            is BayitResult.Loading -> Unit
        }
    }
}

fun ChessViewModel.tapSquare(row: Int, col: Int) {
    val current = _uiState.value as? ChessUiState.GameActive ?: return
    val piece = current.board.getOrNull(row)?.getOrNull(col)

    if (current.selectedSquare != null) {
        if (current.selectedSquare.first == row && current.selectedSquare.second == col) {
            _uiState.value = current.copy(selectedSquare = null)
            return
        }
        if (piece != null && isOwnPiece(piece, current.currentTurn)) {
            _uiState.value = current.copy(selectedSquare = Pair(row, col))
            return
        }
        val from = squareNotation(current.selectedSquare.first, current.selectedSquare.second)
        val to = squareNotation(row, col)
        sendMove(current.game.gameCode, from, to)
        _uiState.value = current.copy(selectedSquare = null)
    } else if (piece != null && isOwnPiece(piece, current.currentTurn)) {
        _uiState.value = current.copy(selectedSquare = Pair(row, col))
    }
}

fun ChessViewModel.sendMove(gameCode: String, from: String, to: String) {
    val moveMsg = """{"type":"move","from":"$from","to":"$to"}"""
    val sent = chessWebSocketHandler.send(moveMsg)
    logger.info("Move sent", mapOf("from" to from, "to" to to, "ws" to sent.toString()))
    if (!sent) {
        launchInScope {
            when (val result = chessRepository.makeMove(gameCode, from, to)) {
                is BayitResult.Success -> {
                    applyGameUpdate(result.data)
                    pollBotReply(gameCode)
                }
                is BayitResult.Error -> logger.error("REST move fallback failed", result.exception)
                is BayitResult.Loading -> Unit
            }
        }
    }
}

fun ChessViewModel.resign(gameCode: String) {
    launchInScope {
        chessWebSocketHandler.send("""{"type":"resign"}""")
        logger.info("Resigned game", mapOf("gameCode" to gameCode))
        when (val result = chessRepository.resignGame(gameCode)) {
            is BayitResult.Success -> applyGameUpdate(result.data)
            is BayitResult.Error -> logger.error("Resign REST call failed", result.exception)
            is BayitResult.Loading -> Unit
        }
    }
}

fun ChessViewModel.offerDraw(gameCode: String) {
    chessWebSocketHandler.send("""{"type":"offer_draw"}""")
    logger.info("Draw offered", mapOf("gameCode" to gameCode))
}

fun ChessViewModel.respondToDraw(accept: Boolean, gameCode: String) {
    chessWebSocketHandler.send("""{"type":"draw_response","accept":$accept}""")
    logger.info("Draw response sent", mapOf("accept" to accept.toString()))
}

fun ChessViewModel.sendChatMessage(gameCode: String, message: String) {
    val wsMsg = """{"type":"chat","message":"${message.replace("\"", "\\\"")}"}"""
    val sent = chessWebSocketHandler.send(wsMsg)
    if (!sent) {
        launchInScope {
            when (val result = chessRepository.sendChatMessage(gameCode, message)) {
                is BayitResult.Success -> {
                    val current = _uiState.value as? ChessUiState.GameActive ?: return@launchInScope
                    _uiState.value = current.copy(
                        chatMessages = current.chatMessages + result.data
                    )
                }
                is BayitResult.Error -> logger.error("Send chat failed", result.exception)
                is BayitResult.Loading -> Unit
            }
        }
    }
}

fun ChessViewModel.toggleChatExpanded() {
    val current = _uiState.value as? ChessUiState.GameActive ?: return
    _uiState.value = current.copy(isChatExpanded = !current.isChatExpanded)
}

fun ChessViewModel.navigateToLobby() {
    wsJob?.cancel()
    timerJob?.cancel()
    chessWebSocketHandler.disconnect()
    _uiState.value = ChessUiState.Lobby()
}

fun ChessViewModel.squareNotation(row: Int, col: Int): String {
    val file = ('a' + col).toString()
    val rank = (8 - row).toString()
    return "$file$rank"
}

internal fun isOwnPiece(piece: Char, currentTurn: String): Boolean {
    return if (currentTurn == "white") piece.isUpperCase() else piece.isLowerCase()
}

File: ios-app/BayitPlusApp/ViewModels/ChessViewModel+Actions.swift
import BayitLocalization
import Foundation

/// Game action methods extracted from ChessViewModel for the 200-line limit.
extension ChessViewModel {
    @MainActor
    func createGameForWhatsApp(
        color: String,
        timeControl: Int?,
        localization: LocalizationManager
    ) async {
        isLoading = true
        error = nil
        do {
            let created = try await repository.createGame(
                color: color, gameMode: "pvp", botDifficulty: nil, timeControl: timeControl
            )
            applyGameState(created)
            await connectWebSocket(gameCode: created.gameCode)
            let link = "https://\(webHost)/chess/\(created.gameCode)"
            let message = localization.t(
                "chess.whatsAppMessage",
                ["code": created.gameCode, "link": link]
            )
            pendingWhatsAppMessage = message
            logger.info("WhatsApp challenge created", context: ["gameCode": created.gameCode])
        } catch {
            if let message = error.userFriendlyMessage {
                self.error = message
            }
            logger.error("Failed to create WhatsApp challenge", error: error)
        }
        isLoading = false
    }

    @MainActor
    func loadGame(gameId: String) async {
        isLoading = true
        error = nil
        do {
            let fetched = try await repository.getGameState(gameId: gameId)
            applyGameState(fetched)
            logger.info("Game loaded", context: ["gameId": gameId])
        } catch {
            if let message = error.userFriendlyMessage {
                self.error = message
            }
            logger.error("Failed to load game", error: error)
        }
        isLoading = false
    }

    @MainActor
    func createGame(color: String, gameMode: String, botDifficulty: String?, timeControl: Int? = nil) async {
        isLoading = true
        error = nil
        do {
            let created = try await repository.createGame(
                color: color, gameMode: gameMode, botDifficulty: botDifficulty, timeControl: timeControl
            )
            applyGameState(created)
            await connectWebSocket(gameCode: created.gameCode)
            logger.info("Game created", context: ["gameCode": created.gameCode])
        } catch {
            if let message = error.userFriendlyMessage {
                self.error = message
            }
            logger.error("Failed to create game", error: error)
        }
        isLoading = false
    }

    @MainActor
    func joinGame(code: String) async {
        isLoading = true
        error = nil
        do {
            let joined = try await repository.joinGame(gameCode: code)
            applyGameState(joined)
            await connectWebSocket(gameCode: joined.gameCode)
            showingJoinSheet = false
            joinCode = ""
            logger.info("Game joined", context: ["gameCode": joined.gameCode])
        } catch {
            if let message = error.userFriendlyMessage {
                self.error = message
            }
            logger.error("Failed to join game", error: error)
        }
        isLoading = false
    }

    @MainActor
    func sendMove(from: (Int, Int), to: (Int, Int)) async {
        let fromNotation = squareNotation(row: from.0, col: from.1)
        let toNotation = squareNotation(row: to.0, col: to.1)
        lastMove = (from: (row: from.0, col: from.1), to: (row: to.0, col: to.1))

        if let conn = connection, await conn.state == .connected {
            let payload = "{\"type\":\"move\",\"from\":\"\(fromNotation)\",\"to\":\"\(toNotation)\"}"
            await sendWSPayload(payload)
        } else if let gameCode = game?.gameCode {
            await sendMoveViaREST(gameCode: gameCode, from: fromNotation, to: toNotation)
        }
        selectedSquare = nil
    }

    @MainActor
    func sendMoveViaREST(gameCode: String, from: String, to: String) async {
        do {
            let updated = try await repository.makeMove(gameCode: gameCode, from: from, to: to)
            applyGameState(updated)
            logger.info("Move via REST", context: ["from": from, "to": to])
            if updated.gameMode == .bot, updated.status == .active {
                Task {
                    try? await Task.sleep(nanoseconds: 1_500_000_000)
                    await pollGameState(gameCode: gameCode)
                }
            }
        } catch {
            logger.warning("Illegal move rejected", context: ["from": from, "to": to])
            lastMove = nil
            selectedSquare = nil
        }
    }

    @MainActor
    func pollGameState(gameCode: String) async {
        do {
            let refreshed = try await repository.getGameState(gameId: gameCode)
            applyGameState(refreshed)
        } catch {
            logger.error("Poll game state failed", error: error)
        }
    }

    @MainActor func resign() async {
        await sendWSPayload("{\"type\":\"resign\"}")
    }

    @MainActor func offerDraw() async {
        await sendWSPayload("{\"type\":\"offer_draw\"}")
        drawOffered = true
    }

    @MainActor func respondToDraw(accept: Bool) async {
        await sendWSPayload("{\"type\":\"draw_response\",\"accept\":\(accept)}")
        drawOffered = false
    }

    @MainActor
    func inviteFriend(friendUserId: String, color: String = "white", timeControl: Int? = nil) async {
        isLoading = true
        error = nil
        do {
            let invited = try await repository.invitePlayer(
                friendUserId: friendUserId, color: color, timeControl: timeControl
            )
            applyGameState(invited)
            await connectWebSocket(gameCode: invited.gameCode)
            logger.info("Friend invited", context: ["friendUserId": friendUserId])
        } catch {
            if let message = error.userFriendlyMessage {
                self.error = message
            }
            logger.error("Failed to invite friend", error: error)
        }
        isLoading = false
    }

    @MainActor
    func sendChatMessage(_ text: String) async {
        if let conn = connection, await conn.state == .connected {
            let escaped = text.replacingOccurrences(of: "\"", with: "\\\"")
            await sendWSPayload("{\"type\":\"chat\",\"message\":\"\(escaped)\"}")
        } else if let gameCode = game?.gameCode {
            await sendChatViaREST(gameCode: gameCode, message: text)
        }
    }

    @MainActor
    private func sendChatViaREST(gameCode: String, message: String) async {
        do {
            let msg = try await repository.sendChatMessage(gameCode: gameCode, message: message)
            chatMessages.append(msg)
        } catch {
            logger.error("REST chat send failed", error: error)
        }
    }
}
